File: app/assets/javascripts/cqm/CQLStatementDependency.js
const mongoose = require('mongoose/browser');

const StatementReferenceSchema = new mongoose.Schema({
  library_name: String,
  statement_name: String,
  hqmf_id: String,
});

const StatementDependencySchema = new mongoose.Schema({
  statement_name: String,
  statement_references: [StatementReferenceSchema],
});

module.exports.StatementReferenceSchema = StatementReferenceSchema;
class StatementReference extends mongoose.Document {
  constructor(object) {
    super(object, StatementReferenceSchema);
  }
}
module.exports.StatementReference = StatementReference;

module.exports.StatementDependencySchema = StatementDependencySchema;
class StatementDependency extends mongoose.Document {
  constructor(object) {
    super(object, StatementDependencySchema);
  }
}
module.exports.StatementDependency = StatementDependency;

File: app/models/cqm/cql_statement_dependency.rb
module CQM
  # Statement is a given CQL Statement within a CQL Library, which has many dependencies
  class StatementDependency
    include Mongoid::Document
    embedded_in :cql_library

    field :statement_name, type: String
    embeds_many :statement_references
  end

  # StatementReference notes a CQL Library/Statement which a given Statement (in which it is embedded) relies upon. This
  # is also used for populations/stratifications/SDEs to reference their defining statement.
  class StatementReference
    include Mongoid::Document
    embedded_in :statement_dependency
    embedded_in :population_set
    embedded_in :population_map
    embedded_in :stratification

    field :library_name, type: String
    field :statement_name, type: String
    field :hqmf_id, type: String
  end
end
